libs/content/workspace/src/workspace.rs

```diff
@@ -83,49 +84,100 @@ pub struct SaveResult {
 }
 
 #[derive(Clone)]
-pub struct WsConfig {
-    pub data_dir: String,
+pub struct WsPersistentStore {
+    pub path: String,
+    pub data: Arc<RwLock<WsPresistentData>>,
+}
 
-    pub auto_save: Arc<AtomicBool>,
-    pub auto_sync: Arc<AtomicBool>,
-    pub zen_mode: Arc<AtomicBool>,
+#[derive(Clone, Serialize, Deserialize, Eq, PartialEq)]
+pub struct WsPresistentData {
+    pub last_open_tabs: Vec<Uuid>,
+    pub auto_save: bool,
+    pub auto_sync: bool,
 }
 
-impl Default for WsConfig {
+impl Default for WsPresistentData {
     fn default() -> Self {
-        Self {
-            data_dir: "".to_string(), // todo: potentially a bad idea
-            auto_save: Arc::new(AtomicBool::new(true)),
-            auto_sync: Arc::new(AtomicBool::new(true)),
-            zen_mode: Arc::new(AtomicBool::new(false)),
-        }
+        Self { auto_save: true, auto_sync: true, last_open_tabs: Vec::default() }
     }
 }
 
-impl WsConfig {
-    pub fn new(dir: String, auto_save: bool, auto_sync: bool, zen_mode: bool) -> Self {
-        let mut s = Self { data_dir: dir, ..Default::default() };
-        s.update(auto_save, auto_sync, zen_mode);
-        s
+impl WsPersistentStore {
+    pub fn new(path: PathBuf) -> Self {
+        match fs::File::open(&path) {
+            Ok(f) => WsPersistentStore {
+                path: path.to_string_lossy().to_string(),
+                data: Arc::new(RwLock::new(serde_json::from_reader(f).unwrap_or_default())),
+            },
+            Err(_) => WsPersistentStore {
+                path: path.to_string_lossy().to_string(),
+                data: Arc::new(RwLock::new(WsPresistentData::default())),
+            },
+        }
     }
 
-    pub fn update(&mut self, auto_save: bool, auto_sync: bool, zen_mode: bool) {
-        self.auto_save.store(auto_save, Ordering::Relaxed);
-        self.auto_sync.store(auto_sync, Ordering::Relaxed);
-        self.zen_mode.store(zen_mode, Ordering::Relaxed);
+    pub fn update<F>(&mut self, updater: F)
+    where
+        F: FnOnce(&mut WsPresistentData),
+    {
+        let mut data_lock = self.data.write().unwrap();
+        let old_data = data_lock.clone();
+
+        updater(&mut data_lock);
```

Comment:
I'm not convinced of the closure here. I think there's a number of simpler paths that won't involve a closure. I can be more explicit if you'd like.